File: leetcode/1345-jump-game-iv/solution.py
from collections import defaultdict, deque
from typing import List


class Solution:
    def minJumps(self, arr: List[int]) -> int:
        n = len(arr)
        if n == 1:
            return 0

        value_to_indices = defaultdict(list)
        for i, v in enumerate(arr):
            value_to_indices[v].append(i)

        visited = [False] * n
        visited[0] = True
        queue = deque([0])
        steps = 0

        while queue:
            for _ in range(len(queue)):
                i = queue.popleft()
                if i == n - 1:
                    return steps

                neighbors = value_to_indices[arr[i]] + [i - 1, i + 1]
                value_to_indices[arr[i]] = []  # consume the whole group once

                for j in neighbors:
                    if 0 <= j < n and not visited[j]:
                        visited[j] = True
                        queue.append(j)
            steps += 1

        return -1

File: leetcode/1345-jump-game-iv/solution.cpp
#include <vector>
#include <unordered_map>
#include <queue>
using namespace std;

class Solution {
public:
    int minJumps(vector<int>& arr) {
        int n = arr.size();
        if (n == 1) {
            return 0;
        }

        unordered_map<int, vector<int>> valueToIndices;
        for (int i = 0; i < n; i++) {
            valueToIndices[arr[i]].push_back(i);
        }

        vector<bool> visited(n, false);
        visited[0] = true;
        queue<int> q;
        q.push(0);
        int steps = 0;

        while (!q.empty()) {
            int size = q.size();
            for (int k = 0; k < size; k++) {
                int i = q.front();
                q.pop();
                if (i == n - 1) {
                    return steps;
                }

                vector<int> neighbors = valueToIndices[arr[i]];
                neighbors.push_back(i - 1);
                neighbors.push_back(i + 1);
                valueToIndices[arr[i]].clear();

                for (int j : neighbors) {
                    if (j >= 0 && j < n && !visited[j]) {
                        visited[j] = true;
                        q.push(j);
                    }
                }
            }
            steps++;
        }

        return -1;
    }
};
